Source/Scene/ModelExperimental/GeometryPipelineStage.js

```diff
@@ -0,0 +1,121 @@
+import defined from "../../Core/defined.js";
+import PrimitiveType from "../../Core/PrimitiveType.js";
+import AttributeType from "../AttributeType.js";
+import VertexAttributeSemantic from "../VertexAttributeSemantic.js";
+import GeometryVS from "../../Shaders/ModelExperimental/GeometryVS.js";
+
+/**
+ * The geometry pipeline stage processes the vertex attributes of a primitive.
+ *
+ * @namespace
+ *
+ * @private
+ */
+var GeometryPipelineStage = {};
+
+/**
+ * This pipeline stage processes the vertex attributes of a mesh primitive, adding the attribute declarations to the shaders,
+ * the attribute objects to the render resources and setting the flags as needed.
+ *
+ * Processes a mesh primitive. This stage modifies the following parts of the render resources:
+ * <ul>
+ *  <li> adds attribute and varying declarations for the vertex attributes in the vertex and fragment shaders
+ *  <li> creates the objects required to create VertexArrays
+ *  <li> sets the flag for point primitive types
+ * </ul>
+ *
+ * @param {MeshPrimitiveRenderResources} renderResources The render resources for this mesh primitive.
+ * @param {ModelComponents.Primitive} primitive The mesh primitive.
+ *
+ * @private
+ */
+GeometryPipelineStage.process = function (renderResources, primitive) {
+  var attributeIndex = 0;
+  var index;
+  for (var i = 0; i < primitive.attributes.length; i++) {
+    var attribute = primitive.attributes[i];
+    if (attribute.semantic !== VertexAttributeSemantic.POSITION) {
+      attributeIndex = attributeIndex + 1;
+      index = attributeIndex;
+    } else {
+      index = 0;
+    }
+    processAttribute(renderResources, attribute, index);
+  }
+
+  var shaderBuilder = renderResources.shaderBuilder;
+  if (primitive.primitiveType === PrimitiveType.POINTS) {
+    shaderBuilder.addDefine("PRIMITIVE_TYPE_POINTS");
+  }
+
+  shaderBuilder.addVertexLines([GeometryVS]);
+  shaderBuilder.addVarying("vec3", "v_positionEC");
+};
+
+function processAttribute(renderResources, attribute, attributeIndex) {
+  var semantic = attribute.semantic;
+  var setIndex = attribute.setIndex;
+  var attributeType = attribute.type;
+
+  var shaderBuilder = renderResources.shaderBuilder;
+
+  var variableName;
+  var varyingName;
+  var glslType = AttributeType.getGlslType(attributeType);
+
+  if (defined(semantic)) {
+    variableName = VertexAttributeSemantic.getVariableName(semantic, setIndex);
+    varyingName = "v_" + variableName;
+
+    shaderBuilder.addVarying(glslType, varyingName);
```

Comment:
In Slack, we discussed this a bit, we want to do two things for now:

1. Always add varyings to the shader even for custom shaders
2. also generate a GLSL function to assign `v_customAttribute = a_customAttribute` and call it in the VS shader

eventually we need a find a way to disable some of this code if not actually used in the shader. But this will be handled in future PRs.

Reply:
This has now been added. It's not a function, just a simple list of statements.